keps/sig-apps/20200302-surge-on-daemonsets.md

```diff
@@ -0,0 +1,163 @@
+
+In the event a rollout cannot proceed due to hitting the MaxSurge limit (due to any condition, whether scheduling, new pods not going ready) the controller should pause creating new pods until conditions change.
+
+DaemonSet pods are slightly more constrained than Deployments when it comes to scheduling issues since each pod is tied to a single node, so it is worth describing exactly how surge pods that violate same node constraints would be handled consistent with Deployments. The most common conflict is use of HostPort within the pod spec across two versions, which would prevent the second pod from landing and the rollout from proceeding. An identical failure would occur with a Deployment of scale 4 on a 3 node cluster - the rollout would be prohibited because the fourth pod could not be scheduled, and so should be handled identically by this controller. It is user error to specify impossible scheduling constraints, and the correct way to convey that is via status conditions on the DaemonSet (which is a separate proposal).
+
+As an optimization for users we can consider rejecting via validation changes that would prevent scheduling of surging daemonsets:
+
+* HostPort use when MaxSurge != 0
+
+However in general we do not perform this validation in other contexts, and its novelty probably does not justify the extra complexity.
+
+### Workload Implications
+
+There are three main workload types that seek to minimize disruption:
+
+1. Infrastructure that should be quickly replaced during update (CNI plugins, CSI plugins).
+2. Infrastructure that wishes to hand off a node resource during an upgrade (socket, namespace, process)
+3. Infrastructure that must remain 100% available to support workloads (networking components, proxies).
+
+In general, all of these benefit from minimizing the time between old pod shutting down and new pod starting up. MaxSurge allows components to arbitrarily approach zero disruption by careful tuning of their launch scripts and access to shared resources, such as sockets or shared disk.
+
+Infrastructure invoked by Kubernetes components (CRI, CNI, CSI) can usually fall within the first category and may require some coordination from the invoking process to minimize downtime. For instance, the Kubelet may retry certain types of CSI errors transparently to mitigate brief disruption to a CSI plugin. Or the container runtime may retry certain CNI errors if the plugin is not available.
+
+The second category of workload requires some coordination between the old and new container - for instance, reusing a host volume and checking for file locking on shared resources, or using the SO_REUSEPORT option to start listening on an interface and share old and new traffic. In general the workload author is assumed to understand how to minimize disruption and Kubernetes is only giving them an overlapping window of execution before beginning the termination of the old process.
+
```

Comment:
we should allow the user controlling the overlap window  between old and new pods. Would that be through readiness probe ?

Reply:
Yes, ready controls gating.  Updated.